cowbird/handlers/impl/geoserver.py

```diff
@@ -245,16 +434,59 @@ def validate_shapefile(self, workspace_name, shapefile_name):
         """
         # Small wait time to prevent unnecessary failure since shapefile is a multi file format
         sleep(1)
-        shapefile_extensions = [".prj", ".dbf", ".shx"]
         files_to_find = [
-            f"{self._shapefile_folder_dir(workspace_name)}/{shapefile_name}{ext}" for ext in shapefile_extensions
+            f"{self._shapefile_folder_dir(workspace_name)}/{shapefile_name}{ext}" for ext in SHAPEFILE_ALL_EXTENSIONS
         ]
         for file in files_to_find:
             if not os.path.isfile(file):
                 LOGGER.warning("Shapefile is incomplete: Missing [%s]", file)
                 raise FileNotFoundError
         LOGGER.info("Shapefile [%s] is valid", shapefile_name)
 
+    def _get_shapefile_permissions(self, workspace_name, shapefile_name):
+        # type:(str, str) -> (bool, bool)
+        """
+        Resolves the actual shapefile permissions on the file system, by checking the permissions of its parent
+        directory and by keeping the less restrictive permissions of the files associated with the shapefile.
+        """
+        datastore_dir_path = self._shapefile_folder_dir(workspace_name)
+        workspace_status = os.stat(datastore_dir_path)[stat.ST_MODE]
+        is_workspace_readable = bool(workspace_status & stat.S_IRUSR and workspace_status & stat.S_IXUSR)
+
+        is_shapefile_readable = False
+        is_shapefile_writable = False
+
+        # If the parent directory is not readable, the user doesn't have read/write access to any contained files.
+        if is_workspace_readable:
+            files_to_check = [
+                f"{datastore_dir_path}/{shapefile_name}{ext}" for ext in SHAPEFILE_ALL_EXTENSIONS
+            ]
+            # The files should normally all have the same permissions, but if any of them has a write/read permission,
+            # we will consider them all as writable/readable in the Magpie resource.
```

Comment:
Yes, currently, it just checks if any of the files of a shapefile has a read or a write permission and applies them to all the shapefile's files if a read/write permission was found. So in your example, the second file with a `rw` permission would indeed undo the previously modified permission.
The current solution also produces the behavior where if you change the permissions of only one file from `rw` to only `r`, the file will get turned back to `rw` since other files of the shapefile will still have the previous `rw` permission.

In any case, I think we definitely have to synchronize them, else some files can become inconsistent with Magpie's permissions.

If we try checking the modified date, I guess we would just prioritize the oldest modified file's permissions?
- in the `on_created` case, the other files will always have an older or equal modified time and would get priority. So in your example, we would synchronize the files and retain the update that used the `r--` permission, since the first file would have the oldest modified time.
- In the `on_modified` case, I'm not sure it would make sense to apply the same logic.  Since in this case, the files would all be already existing, it would make more sense to keep the permissions from the latest modified file.

So, only in the `on_created` case, we could check for the oldest modified files to determine the permissions to apply to all the other files of the shapefile? And for `on_modified`, we just overwrite all the files with the latest modified file's permissions? What do you think?

Reply:
On a similar subject, do you it is a problem if there are different possibilities of Magpie permissions for a file with 
`r--` for example?

What I mean is :
- in some cases, a file can be `chmod` to `0o400` which can result and the corresponding Magpie resource to receive all possible `read` permissions.
- in some other cases, it is a `permission_created` event on a single `read` permission on Magpie that changes the file permissions to `0o400`

In both cases, the file has `0o400` but the file can have different permissions on Magpie.

Do we keep this behavior or would it be better in the case of a `permission_created` to trigger a kind of synchronize event which would add all the other possible read permissions to the Magpie resource? Actually if the `chmod` applied from a `permission_created` event actually triggers an `on_modified` event from the file modification, it might already do this...

I actually need to check what actually happens right now, because I have just tested on `birdhouse`, and a `chmod` doesn't seem to trigger the `on_modified` event. It is detected by the `catalog`, but I don't think the `on_modified` function from Geoserver is called right now.

Reply:
One thing to consider would be which permissions GeoServer applies by default when generating the files. If the `.shp` was generated first and modified with `r--`, I think that the permission should remain `read` even if a `.shx` was generated afterwards with `rw-` by default. So maybe not consider the latest update date-time, but rather the "main" file only?

For the mapping of read permissions, indeed, you would need to check if the `on_modified` results in the following ping-pong sync of permissions. 

1. magpie `create` `getfeature` ⇾ cowbird `chmod` `0o400`
2. cowbird `on_modified` triggered ⇾ magpie `create` `[<perm>]` *for each permission that defines `read`-like behaviour*
3. magpie `create` `[<perm>]` *for each previous generated* ⇾ cowbird `chmod` `0o400`

And the same in the reverse order if a permission operation was `remove`.

If this does not happen automatically from changes detected by the file-system monitoring, then it must be added explicitly.

One important aspect to consider though would be that adding a `read`-like permission should not enforce `r--` (ie: removing `w`). It must only update the `r` part of the file permissions. This is important in this kind of situation where multiple magpie permissions are mapped to `r`, since a `write`-like permission could be added while `magpie`⇿`cowbird` sync for the first `read`-like permissions are still being synchronized.

Reply:
> So maybe not consider the latest update date-time, but rather the "main" file only?

I think it could be another option, which actually goes back to the previous version of code, where only the main `.shp` file extension was processed in the `on_created()` and `on_deleted()` events.

`on_created()`:
- if `.shp` : 
  - we check the permissions and create the associated Magpie permissions
  - update the other files of the shapefile with the same permissions, if they exist
- else (other extensions) : 
  - if there is an existing `.shp` file :
    - apply the same permissions as that file. 
    - create associated Magpie permissions (in case they were not already created by the .shp)
  - if the `.shp` doesn't exist, we do nothing. The Magpie resource will be created when the `.shp` is created.
    
`on_modified()`:
- if `.shp` : update the Magpie permissions accordingly, and update the other files of the shapefile
- else (other extensions) : it would be weird to ignore those and to just put back directly the `.shp` permissions. I think we should do the same as for the `.shp` case here, meaning updating the Magpie permissions (if they exist) and the other files according to the new permissions

`on_deleted()`:
- if `.shp`: delete all other files and the Magpie resource
- else (other extensions) : same thing as the `.shp`?

In summary, for `on_modified()` and `on_deleted()`, all files are treated the same. For `on_created()`, `.shp` has the priority for deciding the permissions.

What do you think?

Reply:
I think this is reasonable with my (limited) understanding of each of those files for the `on_created` case.
I think other files than `.shp` are only in support of the shapefile, or an extra temporary/format file of some sort.
@francisPLT would know better for sure.

If the other files are irrelevant when `.shp` is missing (to be confirmed), I think we could go as far as wiping everything (both on Magpie and GeoServer side) if it is missing. In that case, the proposed `on_deleted` handling would make sense as well.

For the `on_modified`, I'm not sure about the "else". Maybe some files (eg: `.dbf`) will be expected to be writable even if the shapefile is read-only. I'm not sure if GeoServer does any `chmod` or generates any of those files on its own. If that was the case, it would incorrectly override the `.shp` access on Magpie. I think it is safer to assume that if any actual shapefile access must be modified, they should align with the `.shp`, and therefore that file would be `chmod` as well as needed.

I found a few other file extensions to consider, along their description, here: 
https://mxd.codes/articles/what-is-a-shapefile-shp-dbf-and-shx

Reply:
Shapefiles... my favourite archaic format.

While there are separate entities, a shapefile can be considered a single file for permission purposes. Practically speaking, a shapefile is a set of files, each with it's use, like different keys in a Python dictionary.

In our case, Shapefiles will be either imported by the user directly to his workspace, or created by an algorithm or process. Once in the workspace (the shapefile directory of the user) the Shapefile is automatically added as a layer to the Geoserver instance linked to Cowbird, via the automation of the Geoserver handler.

Geoserver could, in theory, require write access for transactional WFS requests, but it's not covered by the use cases. Usually, the only thing Geoserver will do to the Shapefile is read it and create a `<shapefile_name>.qix` file. Actual manipulation of the shapefile would likely be via other processes or tools, but in our context, it would probably be by overwriting the file.

I don't think we'll have a service enabling shapefile manipulation (transactional WFS, QGIS, or some sort of web map editing tool, at least for now), and requirements will likely depend on the actual service.

Usually, individual parts of a Shapefile are not used independently; exceptions are usually with `*.dbf` when you somehow corrupt or break the database part.

As for the deletion part... I would only consider the `delete all if .shp is deleted`, as some of the parts could be removed by a user for other reasons (mainly the optional ones). 

At the very least, I would not expect that deleting, say, the `*.dbf`, would cause the other parts of a shapefile to be deleted automatically, but I would expect that if the Shapefile is removed via another service, that all parts would be removed.

Reply:
I agree with this. Let's use this interpretation of shapefiles as the "main" component to start with, not worrying too much about the other files other than adding the permissions if they were missing. If anything happens to be problematic with these other files later on, we'll address it accordingly then. I also don't think we'll actually have many `write` use cases other than the user modifying the file by themselves via the workspace. We don't need to worry too much about the other files if they are not perfectly in sync.

Reply:
Thanks for the clarifications.
So from what I understand, I just have to process `on_created`, `on_modified` and `on_deleted` events if it concerns the main `.shp` file. For other extensions, I will ignore the event. We will adjust this if required in the future.